problem-gson/src/test/java/org/zalando/problem/GsonProblemTest.java

```diff
@@ -0,0 +1,332 @@
+package org.zalando.problem;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.stream.JsonReader;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.net.URI;
+import java.net.URL;
+import java.util.List;
+import java.util.Objects;
+
+import static com.jayway.jsonassert.JsonAssert.with;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasEntry;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.hasToString;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+import static org.hamcrest.Matchers.startsWith;
+import static org.hobsoft.hamcrest.compose.ComposeMatchers.hasFeature;
+import static org.zalando.problem.Status.BAD_REQUEST;
+
+class GsonProblemTest {
+
+    private static Gson gson;
+
+    @BeforeAll
+    @VisibleForTesting
+    static void init() {
+        GsonBuilder builder = new GsonBuilder();
+        // Register known sub-types
+        ProblemAdapterFactory problemAdapter = new ProblemAdapterFactory()
+                .registerSubtype(InsufficientFundsProblem.class, InsufficientFundsProblem.TYPE)
+                .registerSubtype(OutOfStockException.class, OutOfStockException.TYPE);
+        builder.registerTypeAdapterFactory(problemAdapter);
+        gson = builder.create();
+    }
+
+    private static JsonReader getReader(final String name) throws FileNotFoundException {
+        final ClassLoader loader = Thread.currentThread().getContextClassLoader();
+        URL resource = Objects.requireNonNull(loader.getResource(name), () -> "resource " + name + " not found.");
+        return new JsonReader(new FileReader(new File(resource.getPath())));
+    }
+
+    private String getStackTrace(final Throwable throwable) {
+        final StringWriter writer = new StringWriter();
+        throwable.printStackTrace(new PrintWriter(writer));
+        return writer.toString();
+    }
+
+    @Test
+    @VisibleForTesting
```

Comment:
Isn't that annotation meant to be put on production code to document that it's protected/package-private/public for testing purposes?

Reply:
Since I've SonarLint setup in my IDE, if I make the method public it reports a warning and asks to change the method access to package-private. While Codacy reports error for package-private access modifiers for any field or method. Since I was not sure about Codacy's behavior for public, hence tried resolving this with VisibleForTesting annotation.

I'll try and make the test methods public, getting rid of this annotation. 👍

Reply:
We're using JUnit 5, so package private test methods is correct. Codacy should also be fine with that. You'll need to ignore the warning in your IDE.

Reply:
@whiskeysierra for previous commits Codacy had reported following error for test methods:
> Use explicit scoping instead of the default package private level
To fix this, I had to resort to VisibleForTesting annotation as suggested by Codacy.

This behavior/issue was recently fixed in codacy: https://github.com/codacy/codacy-pmdjava/issues/54, I'm not sure why it was still reporting the error for this PR.

Reply:
I'll disable that rule in Codacy.

Reply:
Done